src/python/merge_tester.py

```diff
@@ -164,96 +186,152 @@ def merge_commits(
-            print(
-                repo_name,
-                merging_method,
-                base,
-                "Exception during testing of the merge. Exception:\n",
-                e,
+            results_merging[merging_method] = (
+                merge_status,
+                runtime,
+                explanation,
+                repo_dir_copy,
             )
 
-    if STORE_SCRATCH:
-        dst_name = os.path.join(
-            SCRATCH_DIR, "_".join([repo_name, left, right, base, merging_method])
+    # Compare the results of the different merging methods.
+    results = {}
+    for idx, merging_method1 in enumerate(MERGE_TOOLS):
+        merge_status1, _, _, path_merge_method1 = results_merging[merging_method1]
+        for merging_method2 in MERGE_TOOLS[(idx + 1) :]:
+            cache_file = os.path.join(
+                cache_dir,
+                "merge_diff_results",
+                merge_id + merging_method1 + "_" + merging_method2,
+            )
+            if os.path.isfile(cache_file + ".txt"):
+                status = read_cache_diff_status(cache_file)
+                results[(merging_method1, merging_method2)] = status
+                continue
+            merge_status2, _, _, path_merge_method2 = results_merging[merging_method2]
+            if (
+                merge_status1 == MERGE_STATES.Merge_success
+                and merge_status2 == MERGE_STATES.Merge_success
+            ):
+                command = f"diff -x .git* -r {path_merge_method1} {path_merge_method2}"
+                with subprocess.Popen(command.split()) as process:
+                    process.wait()
+                    status = process.returncode == 0
+            else:
+                status = False
+            results[(merging_method1, merging_method2)] = status
+            write_cache_diff_status(status, cache_file)
+
+    # Test the merged repos.
+    for merging_method in MERGE_TOOLS:
+        cache_file = os.path.join(
+            cache_dir,
+            "merge_test_results",
+            merge_id + merging_method,
         )
-        if os.path.isdir(dst_name):
-            shutil.rmtree(dst_name, onerror=del_rw)
-        repo_dir_copy_merging_method = os.path.join(repo_dir_copy, merging_method)
-        if os.path.isdir(repo_dir_copy_merging_method):
-            shutil.copytree(repo_dir_copy_merging_method, dst_name)
+        if os.path.isfile(cache_file + ".txt"):
+            status, runtime, _ = read_cache_merge_status(cache_file)
+            results[merging_method] = (status, runtime)
+            continue
+        merge_status, runtime, explanation, repo_dir_copy = results_merging[
+            merging_method
+        ]
+        if merge_status == MERGE_STATES.Merge_success:
+            try:
+                test_status, explanation = repo_test(repo_dir_copy, TIMEOUT_TESTING)
+                if test_status == TEST_STATE.Tests_passed:
+                    merge_status = MERGE_STATES.Tests_passed
```

Comment:
It slightly bugs me that TEST_STATE is singular but MERGE_STATES is plural.  Please be consistent.  (I'm not sure what is considered the best Python style.)